src/Dictionaries/RegExpTreeDictionary.cpp

```diff
@@ -0,0 +1,554 @@
+    auto regex_column = block.getByName(kRegExp).column;
+    auto keys_column = block.getByName(kKeys).column;
+    auto values_column = block.getByName(kValues).column;
+
+    size_t size = block.rows();
+    for (size_t i = 0; i < size; i++)
+    {
+        UInt64 id = id_column->getUInt(i);
+        UInt64 parent_id = pid_column->getUInt(i);
+        String regex = (*regex_column)[i].safeGet<String>();
+
+        if (regex_nodes.contains(id))
+            throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "There are duplicate id {}", id);
+
+        if (id == 0)
+            throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "There are invalid id {}", id);
+
+        regexps.push_back(regex);
+        regexp_ids.push_back(id);
+
+        re2_st::RE2::Options regexp_options;
+        regexp_options.set_log_errors(false);
+        RegexTreeNodePtr node = std::make_unique<RegexTreeNode>(id, parent_id, regex, regexp_options);
+
+        int num_captures = std::min(node->searcher.NumberOfCapturingGroups() + 1, 10);
+
+        Array keys = (*keys_column)[i].safeGet<Array>();
+        Array values = (*values_column)[i].safeGet<Array>();
+        size_t keys_size = keys.size();
+        for (size_t i = 0; i < keys_size; i++)
+        {
+            const String & name = keys[i].safeGet<String>();
+            const String & value = values[i].safeGet<String>();
+            if (structure.hasAttribute(name))
+            {
+                const auto & attr = structure.getAttribute(name);
+                auto string_pieces = createStringPieces(value, num_captures, regex, logger);
+                if (!string_pieces.empty())
+                {
+                    node->attributes[name] = RegexTreeNode::AttributeValue{.field = values[i], .pieces = std::move(string_pieces)};
+                }
+                else
+                {
+                    Field field = parseStringToField(values[i].safeGet<String>(), attr.type);
+                    node->attributes[name] = RegexTreeNode::AttributeValue{.field = std::move(field)};
+                }
+            }
+        }
+        regex_nodes.emplace(id, std::move(node));
+    }
+}
+
+void RegExpTreeDictionary::initGraph()
+{
+    for (const auto & [id, value]: regex_nodes)
+    {
+        UInt64 pid = value->parent_id;
+        if (pid == 0) // this is root
+            continue;
+        if (regex_nodes.contains(pid))
+            regex_nodes[pid]->children.push_back(id);
+        else
+            throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "Unknown parent id {}", pid);
+    }
+    std::set<UInt64> visited;
+    UInt64 topology_id = 0;
+    for (const auto & [id, value]: regex_nodes)
+        if (value->parent_id == 0) // this is root node.
+            initTopologyOrder(id, visited, topology_id);
+    if (topology_order.size() != regex_nodes.size())
+        throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "Invalid Regex tree");
```

Comment:
This exception message is not really informative for a user, could you please update it (or replace with `LOGICAL ERROR` if it's not for a user)? (what is invalid? how to fix it?)

Reply:
actually this belongs to LOGICAL_ERROR. fixing in https://github.com/ClickHouse/ClickHouse/pull/48632
Thank you!